library/Booster/Pattern/ApplyEquations.hs

```diff
@@ -391,34 +391,25 @@ iterateEquations direction preference startTerm = do
             -- the equation evaluation does not change the term any more.
             resetChanged
             -- evaluate functions and simplify (recursively at each level)
-            newTerm <- applyTerm direction preference llvmResult
+            newTerm <-
+                let simp = cached Equations $ traverseTerm direction simp (applyHooksAndEquations preference)
+                 in simp llvmResult
             changeFlag <- getChanged
             if changeFlag
                 then checkForLoop newTerm >> resetChanged >> go newTerm
                 else pure llvmResult
 
 llvmSimplify :: forall io. MonadLoggerIO io => Term -> EquationT io Term
-llvmSimplify t = do
+llvmSimplify term = do
     config <- getConfig
     case config.llvmApi of
-        Nothing -> pure t
+        Nothing -> pure term
         Just api -> do
             logOtherNS "booster" (LevelOther "Simplify") "Calling LLVM simplification"
-            llvmEval config.definition api t
-
-llvmEval :: MonadLoggerIO io => KoreDefinition -> LLVM.API -> Term -> EquationT io Term
-llvmEval definition api = eval
+            let simp = cached LLVM $ evalLlvm config.definition api $ traverseTerm BottomUp simp pure
```

Comment:
`BottomUp` or `TopDown` are equivalent if `onEval` is just `pure`:

```
            | direction == BottomUp -> do
                -- evaluate arguments first
                args' <- mapM onRecurse args
                -- then try to apply equations
                pure $ SymbolApplication sym sorts args'
            | otherwise {- direction == TopDown -} -> do
                -- try to apply equations
                t <- pure app
                SymbolApplication sym sorts
                            <$> mapM onRecurse args
```

Reply:
Indeed they are... but somehow I don't see how the cut-off is implemented.
If we have nested concrete terms, e.g., `+Int(2, +Int(3, 4))` ,Is the llvm backend going to be called twice, for `3 + 4` and then `2 + 7`, or is it called just once as a top-down traversal with cut-off would?